lib/kafka/ec2/mixed_instance_assignment_strategy.rb

```diff
@@ -86,15 +87,19 @@ def assign(members:, topics:)
             end
 
             member_id_to_acceptable_partition_weight[member_id] = acceptable_partition_weight
+            instance_id_to_total_acceptable_partition_weight[instance_id] += acceptable_partition_weight
           end
         end
 
         while last_index < topic_partitions.size
-          member_id, _ = member_id_to_acceptable_partition_weight.max_by { |_, remaining| remaining }
+          instance_id, _ = instance_id_to_total_acceptable_partition_weight.max_by { |_, remaining| remaining }
```

Comment:
I'm concerned that, in the following case, most of the remaining partitions are assigned to the instance of c5.24xlarge:

* One of the instance types is c5.24xlarge (96 consumers) and each member's acceptable partition weight is 1, that is, the total acceptable partition weight is 96.
* One of the instance types is m5.xlarge (4 consumers) and each member's acceptable partition weight is 10, that is, the total acceptable partition weight is 40.

How about the following code?

```diff
diff --git a/lib/kafka/ec2/mixed_instance_assignment_strategy.rb b/lib/kafka/ec2/mixed_instance_assignment_strategy.rb
index 562dc7f..e8d0850 100644
--- a/lib/kafka/ec2/mixed_instance_assignment_strategy.rb
+++ b/lib/kafka/ec2/mixed_instance_assignment_strategy.rb
@@ -41,6 +41,7 @@ module Kafka
         group_assignment = {}
         instance_id_to_capacity = Hash.new(0)
         instance_id_to_member_ids = Hash.new { |h, k| h[k] = [] }
+        member_id_to_instance_id = {}
         total_capacity = 0

         instance_family_to_capacity = @instance_family_weights.is_a?(Proc) ? @instance_family_weights.call() : @instance_family_weights
@@ -51,6 +52,7 @@ module Kafka

           instance_id, instance_type, az = member_id_to_metadata[member_id].split(",")
           instance_id_to_member_ids[instance_id] << member_id
+          member_id_to_instance_id[member_id] = instance_id
           capacity = calculate_capacity(instance_type, az, instance_family_to_capacity, az_to_capacity, weights)
           instance_id_to_capacity[instance_id] += capacity
           total_capacity += capacity
@@ -70,6 +72,7 @@ module Kafka

         last_index = 0
         member_id_to_acceptable_partition_weight = {}
+        instance_id_to_total_acceptable_partition_weight = Hash.new(0)
         instance_id_to_capacity.each do |instance_id, capacity|
           member_ids = instance_id_to_member_ids[instance_id]
           member_ids.each do |member_id|
@@ -86,15 +89,19 @@ module Kafka
             end

             member_id_to_acceptable_partition_weight[member_id] = acceptable_partition_weight
+            instance_id_to_total_acceptable_partition_weight[instance_id] += acceptable_partition_weight
           end
         end

         while last_index < topic_partitions.size
-          member_id, _ = member_id_to_acceptable_partition_weight.max_by { |_, remaining| remaining }
+          max_acceptable_partition_weight = member_id_to_acceptable_partition_weight.values.max
+          member_ids = member_id_to_acceptable_partition_weight.select { |_, w| w == max_acceptable_partition_weight }.keys
+          member_id = member_ids.max_by { |id| instance_id_to_total_acceptable_partition_weight[member_id_to_instance_id[id]] }
           topic, partition = topic_partitions[last_index]
           group_assignment[member_id].assign(topic, [partition])

           member_id_to_acceptable_partition_weight[member_id] -= partition_weights.dig(topic, partition)
+          instance_id_to_total_acceptable_partition_weight[member_id_to_instance_id[member_id]] -= partition_weights.dig(topic, partition)

           last_index += 1
         end
```